### PR TITLE
Add CHIP_Crypto_HMAC implementation

### PR DESCRIPTION
**Pull request by tcarmelveilleux**
#### Problem

- Add CHIP_Crypto_HMAC implementation, which is needed to implement CASE to spec and was missing.

#### Change overview
- CHIP_Crypto_HMAC primitive is needed to compute
  CASE destination ID.
- Added OpenSSL and mbedTLS versions.
- Added basic unit test.

#### Testing
Testing done: unit test pass on OpenSSL and mbedTLS
backends on Linux. Unit tests use test vectors from RFC4231.

Issue #5934

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

File: src/crypto/CHIPCryptoPAL.h [hunk 1]
```diff
@@ -39,26 +39,28 @@ namespace Crypto {
...
 
-// TODO: Consider renaming these values to be closer to definisions in the spec:
-// CHIP_CRYPTO_GROUP_SIZE_BYTES
-// CHIP_CRYPTO_PUBLIC_KEY_SIZE_BYTES
 constexpr size_t kP256_FE_Length                  = 32;
 constexpr size_t kP256_ECDSA_Signature_Length_Raw = (2 * kP256_FE_Length);
 constexpr size_t kP256_Point_Length               = (2 * kP256_FE_Length + 1);
 constexpr size_t kSHA256_Hash_Length              = 32;
 
+constexpr size_t CHIP_CRYPTO_GROUP_SIZE_BYTES      = kP256_FE_Length;
```
Comment (Damian-Nordic) [inline src/crypto/CHIPCryptoPAL.h:47]: I know it's to align the constant naming with the spec, but having two constants with different naming conventions for a single thing looks very confusing. Can we stick to the `kConstant` naming convention and just add annotate the constants with spec names in comments? E.g.
`constexpr size_t kSHA256_Hash_Length              = 32; // CHIP_CRYPTO_HASH_LEN_BYTES in the spec`

**Review by Damian-Nordic — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

File: src/crypto/tests/HMAC_SHA256_test_vectors.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ *    limitations under the License.
+ */
+
+/**
+ * @file - This file contains HMAC SHA256 test vectors.
+ *
+ * See https://tools.ietf.org/html/rfc4231#section-4.7
+ */
+
+#ifndef _HMAC_SHA256_TEST_VECTOR
```
[the remark this answers is not in the document]
Comment (tcarmelveilleux) [reply src/crypto/tests/HMAC_SHA256_test_vectors.h:24]: Just cut'n'pasted the other test vector bodies

File: src/crypto/CHIPCryptoPAL.h (hunk 1 above)
Comment (tcarmelveilleux) [reply src/crypto/CHIPCryptoPAL.h:47]: A lot of the reason code is hard to review for spec compliance is that it uses different names for things like crypto primitives and constants (which are very well defined in spec!) in the implementation, compared to the spec. I can remove this one if it matters a lot, but `kP256_FE_Length` is actually less telling than CHIP_CRYPTO_GROUP_SIZE_BYTES, and similarly, CHIP_CRYPTO_HASH_LEN_BYTES maps to SHA256 today, but may not in the future.

File: src/crypto/tests/HMAC_SHA256_test_vectors.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (tcarmelveilleux) [reply src/crypto/tests/HMAC_SHA256_test_vectors.h:24]: I can do a follow-up PR that cleans them all up, which would be easier and better.

File: src/crypto/CHIPCryptoPAL.h (hunk 1 above)
Comment (bzbarsky-apple) [reply src/crypto/CHIPCryptoPAL.h:47]: @tcarmelveilleux Worth a followup to get rid of kP256_FE_Length?

File: src/crypto/CHIPCryptoPAL.h [hunk 3]
```diff
@@ -430,6 +432,30 @@ class HKDF_sha
...
+     * @brief A function that implements SHA-256 based HMAC per FIPS1981.
+     *
+     * The `out_length` must be at least CHIP_CRYPTO_HASH_LEN_BYTES
+     *
+     * @param key The key to use for the HMAC operation
+     * @param key_length Length of the key
+     * @param message Message over which to compute the HMAC
+     * @param message_length Length of the message over which to compute the HMAC
+     * @param out_buffer Pointer to buffer to write output into.
+     * @param out_length Resulting length of out_buffer
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPAL.h:451]: "Resulting" in what sense?  This is an in param...

Presumably this is just the available length of the buffer, and on success this function writes CHIP_CRYPTO_HASH_LEN_BYTES bytes to the buffer?

File: src/crypto/hsm/CHIPCryptoPALHsm.h [hunk 4]
```diff
@@ -185,6 +185,27 @@ class HKDF_shaHSM : public HKDF_sha
...
 
+#if ENABLE_HSM_HMAC_SHA256
+
+class HMAC_shaHSM : public HMAC_sha
+{
+public:
+    HMAC_shaHSM();
+    ~HMAC_shaHSM();
+
+    virtual CHIP_ERROR HMAC_SHA256(const uint8_t * key, size_t key_length, const uint8_t * message, size_t message_length,
```
Comment (bzbarsky-apple) [inline src/crypto/hsm/CHIPCryptoPALHsm.h:196]: This should have `override`?

**Pull request merged**